### ISSUE TITLE
[Feature Request] Rename stashes

### ISSUE THREAD

**Issue by fbartho** [fork-dev/Tracker#218]
Right-click menu on stash entries allows user to `Apply` a stash or `Delete` a stash, (which I was happy to discover).

It'd be nice if there was a built-in helper that would rename stashes as well.

**Comment by nebhale** [comment]
Great idea since whenever I create stashes at the command line, I **always** forget to give them names.

**Comment by fbartho** [comment]
@nebhale that’s *exactly* my usecase too!

**Comment by DanPristupov** [comment]
I made some research and I could not find any easy way to rename a stash. Any idea how this could be done?

**Comment by fbartho** [comment]
My old hack: Apply, stash save w/ name, delete old stash. 
Note: this makes a fresh stash at the top of the stack, so your stashes get reordered :(

**Comment by DanPristupov** [comment]
@fbartho it also requires user to have clean working directory :(.

There's a hack: https://stackoverflow.com/questions/25931026/how-can-i-rename-a-git-stash#35549615. But it does not really rename stash, it only changes reflog message. In this way new message will be displayed when you call `git stash list`, but all git clients will still show the old title.

**Comment by fbartho** [comment]
Oh! I found `git stash store` https://stackoverflow.com/a/35549615

**Comment by davecotter** [comment]
why is this closed? it is a legit feature request with a defined solution?

**Comment by DanPristupov** [comment]
This is a legit feature request, but I don't know a way to rename a stash. I think currently it's simply not possible.

**Comment by brechtm** [comment]
I think have I figured out a way to do this. I have described it in [this StackOverflow answer](https://stackoverflow.com/a/65627640/438249).

Fork does have an issue with the newly created stash, however. The first character of the stash message is missing. The GitLens VSCode plugin does show the stash message properly in its list of stashes, however.

Related to this, I noticed that Fork hides the `On <branch-name>: ` prefixed to the user-supplied stash message. I assume this is what is causing the missing character. I'm would vote to not remove this however, since this can be quite useful information.

**Comment by DanPristupov** [comment]
@brechtm this looks like a correct approach. I will have a look.

Just a note: a stash can have 3 parents when it's created with the `--keep-index` option.

**Comment by brechtm** [comment]
> Just a note: a stash can have 3 parents when it's created with the `--keep-index` option.

Thanks for the heads-up. I've now mentioned that in the SO answer.

**Comment by DanPristupov** [comment]
Alright, we have a working version! Would you like to try it?

https://cdn.fork.dev/mac/Fork-2.2.3.dmg

The usage is simple, stash rename is available in the context menu in the commit list and on the sidebar.

<img width="321" alt="Screenshot 2021-01-11 at 17 51 23" src="https://user-images.githubusercontent.com/618115/104212543-a0a37980-5435-11eb-836f-8bc937423545.png">

**Comment by fbartho** [comment]
@DanPristupov I was able to test it! and it worked!

Nuance for anybody surprised -- when you rename a stash, it moves to the top of the stack, but that's an acceptable detail imho.

**Comment by brechtm** [comment]
> Alright, we have a working version! Would you like to try it?

Yup, works great!

> Nuance for anybody surprised -- when you rename a stash, it moves to the top of the stack, but that's an acceptable detail imho.

I would also prefer that the ordering to stay the same. If the stashes are sorted by date, the original stash commit's timestamp could perhaps be copied?

**Comment by fbartho** [comment]
@brechtm I believe the stashes are numbered / ordered automatically (not dynamically sorted based on any particular criteria).

It wouldn't surprise me if it were impossible to rename "in-place" without changing the order.

**Comment by brechtm** [comment]
@fbartho Ah yes, I forgot the stashes are part of a stack, and numbered based on their position.

Poking around a bit, I learned that the stash list is basically shoehorned into a reflog. I was able to reorder the stashes by swapping lines of `.git/logs/refs/stash`. In that sense, the `git stash store` step could be replaced by editing the appropriate line of that file: replacing the stash message and commit ID (which is also used in the next entry on the stash stack!).

Editing that file may not be a good idea though. Perhaps it's possible to edit one of the entries in this file using plumbing commands?

In any case, even with the renamed stash jumping to the top, this is still a very useful feature to have as-is. Much appreciated, @DanPristupov! Perhaps in-place renaming (and arbitrary stash reordering?) can be added at a later point in time.
